src/DotVVM.Framework/Runtime/Caching/SimpleLruDictionary.cs

```diff
@@ -0,0 +1,184 @@
+        where TValue: class
+    {
+        // new generation
+        private ConcurrentDictionary<TKey, TValue> hot = new ConcurrentDictionary<TKey, TValue>();
+        // old generation
+        private ConcurrentDictionary<TKey, TValue> cold = new ConcurrentDictionary<TKey, TValue>();
+        // free to take for GC. however, if the GC does not want to collect, we can still use it
+        private readonly ConcurrentDictionary<TKey, WeakReference<TValue>> dead = new ConcurrentDictionary<TKey, WeakReference<TValue>>();
+        private TimeSpan lastCleanupTime = TimeSpan.MinValue;
+
+        private readonly int generationSize;
+        private readonly TimeSpan generationTickTime;
+        private readonly Stopwatch stopWatch = Stopwatch.StartNew(); // this is more reliable source of time information than DateTime.Now (due to system-wide time change, ...)
+
+        public SimpleLruDictionary(int generationSize, TimeSpan generationTickTime)
+        {
+            this.generationSize = generationSize;
+            this.generationTickTime = generationTickTime;
+            SetupTimer();
+        }
+
+        private object locker = new object();
+        internal bool Cleanup(int minSize, TimeSpan minTime)
+        {
+            lock (locker)
+            {
+                // check that's still needed
+                if (minSize > hot.Count) return false;
+                if (minTime > lastCleanupTime) return false;
+
+                lastCleanupTime = stopWatch.Elapsed;
+
+                foreach (var i in cold)
+                {
+                    dead.TryAdd(i.Key, new WeakReference<TValue>(i.Value));
+                }
+                var newHot = new ConcurrentDictionary<TKey, TValue>();
+                cold = hot;
+                hot = newHot;
+            }
+            CleanDeadReferences();
+            return true;
+        }
+
+        private void CleanDeadReferences()
+        {
+            foreach (var i in dead)
+            {
+                if (!i.Value.TryGetTarget(out var _))
+                {
+                    dead.TryRemove(i.Key, out var _);
+                    // at this point, I could have removed different item than what I have checked. But it was marked dead anyway, so it won't really corrupt anything
+                    // (i.e. we won't have two different items, just drop item that should not have been really dropped)
+                }
+            }
+        }
+
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            if (hot.TryGetValue(key, out value))
+                return true;
+            lock (locker)
+            {
+                if (hot.TryGetValue(key, out value))
+                    return true;
+                if (cold.TryGetValue(key, out value))
+                {
+                    // add back to `hot` -- this way, we mark the item as used
+                    hot.TryAdd(key, value);
+                    return true;
```

Comment:
Is it OK that we copy the item to the hot dictionary while keeping it in cold?

Reply:
That does not matter much. Clearing it from `cold` won't free any memory, though, so it's not useful for anything.